src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_TupleBinaryOperator.cs

```diff
@@ -74,7 +74,21 @@ private BoundExpression ReplaceTerminalElementsWithTemps(BoundExpression expr, T
             // Examples:
             // in `expr == (..., ...)` we need to save `expr` because it's not a tuple literal
             // in `(..., expr) == (..., (..., ...))` we need to save `expr` because it is used in a simple comparison
-            return EvaluateSideEffectingArgumentToTemp(VisitExpression(expr), initEffects, ref temps);
+            BoundExpression loweredExpr = VisitExpression(expr);
+            if ((bool)(loweredExpr.Type != null) && NullableAlwaysHasValue((BoundExpression)loweredExpr) is var value && value != null)
+            {
+                // Optimization: if the nullable expression always has a value, we'll save the value to a temp and wrap
+                // that temp so it can be recognized as always having a value later on.
+                var savedValue = EvaluateSideEffectingArgumentToTemp(value, initEffects, ref temps);
+                SyntaxNode syntax = loweredExpr.Syntax;
+                _ = TryGetSpecialTypeMethod(syntax, SpecialMember.System_Nullable_T__ctor, out MethodSymbol nullableCtor);
+
+                return new BoundObjectCreationExpression(syntax, nullableCtor, ImmutableArray.Create(savedValue),
+                    argumentNamesOpt: default, argumentRefKindsOpt: default, expanded: false, argsToParamsOpt: default,
+                    constantValueOpt: null, initializerExpressionOpt: null, binderOpt: null, type: loweredExpr.Type);
+            }
+
+            return EvaluateSideEffectingArgumentToTemp(loweredExpr, initEffects, ref temps);
```

Comment:
>return EvaluateSideEffectingArgumentToTemp(loweredExpr, initEffects, ref temps); [](start = 12, length = 80)

We should also optimize NullableNeverHasValue case, an expression like that doesn't need caching

Reply:
I don't think it needs to be handled specially.
There are two cases:
- an element that never has value; then it has no side effects, and will not be saved to a temp anyways, and the lowering for element-wise binary operator will handle it,
- a tuple that never has value: this case could be further optimized, but the way of getting that scenario is very contrived and the complexity that adds to the code for rewriting nested case does not seem worthwhile to me.

Reply:
> an element that never has value; then it has no side effects, and will not be saved to a temp anyways

It is not clear what makes you think this. Have you tried testing with something other than null or default?

>  a tuple that never has value: this case could be further optimized, but the way of getting that scenario is very contrived and the complexity that adds to the code for rewriting nested case does not seem worthwhile to me.

I don't believe this will add much complexity. Somehow TestEqualOnNullableVsNullableTuples_TupleAlwaysNull screams at me that we should do better. For whatever reason we thought that handling cases like this specially for regular comparisons is worthwhile, I cannot think of any reason why this should be different for tuple comparison. Also, consistency is a good thing. 

---
In reply to: [176383420](https://github.com/dotnet/roslyn/pull/25644#discussion_r176383420) [](ancestors = 176383420)

Reply:
I've tried multiple expressions, such as `new` and casts. The lowered node for elements that never have a value is a bound default expression, which is recognized have having no side effects.

I agree that the generated IL could be simpler (that's why I included that test).
In terms of complexity, I had tried this over the weekend and found it complex (there were repercussions which I don't remember the details of), especially in the context of this rewriting which is already not simply. I'd rather add a comment to mention the "never null" case. 
I can try again, but frankly I feel this is a waste of time, as writing nullable tuples that are always null is contrived (cast or new) in the context of tuple equality.

---
In reply to: [176505498](https://github.com/dotnet/roslyn/pull/25644#discussion_r176505498) [](ancestors = 176505498,176383420)

Reply:
Ok, I think the problem is that I wanted to skip computing the `GetValueOrDefault` portion in that case, but I can leave it in (it will be optimized away anyways).

---
In reply to: [176557640](https://github.com/dotnet/roslyn/pull/25644#discussion_r176557640) [](ancestors = 176557640,176505498,176383420)